File: skill/src/proto.test.ts
import { describe, expect, it } from "vitest";
import { create, toBinary, fromBinary } from "@bufbuild/protobuf";
import {
	EnvelopeSchema,
	EncryptedPayloadSchema,
	PlaintextPayloadSchema,
	HandshakeSchema,
	AuthChallengeSchema,
	AuthResponseSchema,
	MessageType,
	MessageTypeSchema,
} from "@pinch/proto/pinch/v1/envelope_pb.js";

describe("Envelope serialization round-trip", () => {
	it("should round-trip an Envelope with EncryptedPayload", () => {
		const messageId = new Uint8Array([
			1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
		]);
		const senderPubKey = new Uint8Array(32);
		for (let i = 0; i < 32; i++) senderPubKey[i] = i;
		const nonce = new Uint8Array(24);
		for (let i = 0; i < 24; i++) nonce[i] = i + 100;
		const ciphertext = new TextEncoder().encode("encrypted-data-here");

		const original = create(EnvelopeSchema, {
			version: 1,
			fromAddress: "pinch:abc123@relay.example.com",
			toAddress: "pinch:def456@relay.example.com",
			type: MessageType.MESSAGE,
			messageId,
			timestamp: BigInt(Date.now()),
			payload: {
				case: "encrypted",
				value: create(EncryptedPayloadSchema, {
					nonce,
					ciphertext,
					senderPublicKey: senderPubKey,
				}),
			},
		});

		// Serialize
		const data = toBinary(EnvelopeSchema, original);
		expect(data).toBeInstanceOf(Uint8Array);
		expect(data.length).toBeGreaterThan(0);

		// Deserialize
		const decoded = fromBinary(EnvelopeSchema, data);

		// Verify all fields
		expect(decoded.version).toBe(1);
		expect(decoded.fromAddress).toBe("pinch:abc123@relay.example.com");
		expect(decoded.toAddress).toBe("pinch:def456@relay.example.com");
		expect(decoded.type).toBe(MessageType.MESSAGE);
		expect(decoded.messageId).toEqual(messageId);
		expect(decoded.timestamp).toBe(original.timestamp);

		// Verify oneof payload discrimination
		expect(decoded.payload.case).toBe("encrypted");
		if (decoded.payload.case === "encrypted") {
			expect(decoded.payload.value.nonce).toEqual(nonce);
			expect(decoded.payload.value.ciphertext).toEqual(ciphertext);
			expect(decoded.payload.value.senderPublicKey).toEqual(senderPubKey);
		}
	});

	it("should round-trip an Envelope with Handshake payload", () => {
		const signingKey = new Uint8Array(32);
		const encryptionKey = new Uint8Array(32);
		for (let i = 0; i < 32; i++) {
			signingKey[i] = i;
			encryptionKey[i] = i + 32;
		}

		const original = create(EnvelopeSchema, {
			version: 1,
			fromAddress: "pinch:abc123@relay.example.com",
			type: MessageType.HANDSHAKE,
			payload: {
				case: "handshake",
				value: create(HandshakeSchema, {
					version: 1,
					signingKey,
					encryptionKey,
				}),
			},
		});

		const data = toBinary(EnvelopeSchema, original);
		const decoded = fromBinary(EnvelopeSchema, data);

		expect(decoded.payload.case).toBe("handshake");
		if (decoded.payload.case === "handshake") {
			expect(decoded.payload.value.version).toBe(1);
			expect(decoded.payload.value.signingKey).toEqual(signingKey);
			expect(decoded.payload.value.encryptionKey).toEqual(encryptionKey);
		}
	});

	it("should round-trip an Envelope with AuthChallenge payload", () => {
		const nonce = new Uint8Array(32);
		for (let i = 0; i < nonce.length; i++) nonce[i] = i + 11;
		const now = BigInt(Date.now());

		const original = create(EnvelopeSchema, {
			version: 1,
			type: MessageType.AUTH_CHALLENGE,
			timestamp: now,
			payload: {
				case: "authChallenge",
				value: create(AuthChallengeSchema, {
					version: 1,
					nonce,
					issuedAtMs: now,
					expiresAtMs: now + 10_000n,
					relayHost: "relay.example.com",
				}),
			},
		});

		const data = toBinary(EnvelopeSchema, original);
		const decoded = fromBinary(EnvelopeSchema, data);

		expect(decoded.payload.case).toBe("authChallenge");
		if (decoded.payload.case === "authChallenge") {
			expect(decoded.payload.value.version).toBe(1);
			expect(decoded.payload.value.nonce).toEqual(nonce);
			expect(decoded.payload.value.issuedAtMs).toBe(now);
			expect(decoded.payload.value.expiresAtMs).toBe(now + 10_000n);
			expect(decoded.payload.value.relayHost).toBe("relay.example.com");
		}
	});

	it("should round-trip an Envelope with AuthResponse payload", () => {
		const publicKey = new Uint8Array(32);
		const signature = new Uint8Array(64);
		const nonce = new Uint8Array(32);
		for (let i = 0; i < publicKey.length; i++) {
			publicKey[i] = i;
			nonce[i] = 100 + i;
		}
		for (let i = 0; i < signature.length; i++) {
			signature[i] = 200 + (i % 32);
		}

		const original = create(EnvelopeSchema, {
			version: 1,
			type: MessageType.AUTH_RESPONSE,
			payload: {
				case: "authResponse",
				value: create(AuthResponseSchema, {
					version: 1,
					publicKey,
					signature,
					nonce,
				}),
			},
		});

		const data = toBinary(EnvelopeSchema, original);
		const decoded = fromBinary(EnvelopeSchema, data);

		expect(decoded.payload.case).toBe("authResponse");
		if (decoded.payload.case === "authResponse") {
			expect(decoded.payload.value.version).toBe(1);
			expect(decoded.payload.value.publicKey).toEqual(publicKey);
			expect(decoded.payload.value.signature).toEqual(signature);
			expect(decoded.payload.value.nonce).toEqual(nonce);
		}
	});
});

describe("PlaintextPayload serialization round-trip", () => {
	it("should round-trip PlaintextPayload with all fields", () => {
		const content = new TextEncoder().encode("hello");
		const now = BigInt(Date.now());

		const original = create(PlaintextPayloadSchema, {
			version: 1,
			sequence: 42n,
			timestamp: now,
			content,
			contentType: "text/plain",
		});

		const data = toBinary(PlaintextPayloadSchema, original);
		const decoded = fromBinary(PlaintextPayloadSchema, data);

		expect(decoded.version).toBe(1);
		expect(decoded.sequence).toBe(42n);
		expect(decoded.timestamp).toBe(now);
		expect(decoded.content).toEqual(content);
		expect(decoded.contentType).toBe("text/plain");
	});
});

describe("MessageType enum values", () => {
	it("should have correct integer values", () => {
		expect(MessageType.UNSPECIFIED).toBe(0);
		expect(MessageType.HANDSHAKE).toBe(1);
		expect(MessageType.AUTH_CHALLENGE).toBe(2);
		expect(MessageType.AUTH_RESPONSE).toBe(3);
		expect(MessageType.MESSAGE).toBe(4);
		expect(MessageType.DELIVERY_CONFIRM).toBe(5);
		expect(MessageType.CONNECTION_REQUEST).toBe(6);
		expect(MessageType.CONNECTION_RESPONSE).toBe(7);
		expect(MessageType.HEARTBEAT).toBe(8);
	});
});

File: relay/internal/protocol/proto_test.go
package protocol_test

import (
	"testing"
	"time"

	pinchv1 "github.com/pinch-protocol/pinch/gen/go/pinch/v1"
	"google.golang.org/protobuf/proto"
)

func TestEnvelopeRoundTrip(t *testing.T) {
	messageID := []byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
	senderPubKey := make([]byte, 32)
	for i := range senderPubKey {
		senderPubKey[i] = byte(i)
	}
	nonce := make([]byte, 24)
	for i := range nonce {
		nonce[i] = byte(i + 100)
	}
	ciphertext := []byte("encrypted-data-here")

	original := &pinchv1.Envelope{
		Version:     1,
		FromAddress: "pinch:abc123@relay.example.com",
		ToAddress:   "pinch:def456@relay.example.com",
		Type:        pinchv1.MessageType_MESSAGE_TYPE_MESSAGE,
		MessageId:   messageID,
		Timestamp:   time.Now().UnixMilli(),
		Payload: &pinchv1.Envelope_Encrypted{
			Encrypted: &pinchv1.EncryptedPayload{
				Nonce:           nonce,
				Ciphertext:      ciphertext,
				SenderPublicKey: senderPubKey,
			},
		},
	}

	// Serialize
	data, err := proto.Marshal(original)
	if err != nil {
		t.Fatalf("failed to marshal envelope: %v", err)
	}

	// Deserialize
	decoded := &pinchv1.Envelope{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("failed to unmarshal envelope: %v", err)
	}

	// Verify all fields
	if decoded.Version != original.Version {
		t.Errorf("version mismatch: got %d, want %d", decoded.Version, original.Version)
	}
	if decoded.FromAddress != original.FromAddress {
		t.Errorf("from_address mismatch: got %q, want %q", decoded.FromAddress, original.FromAddress)
	}
	if decoded.ToAddress != original.ToAddress {
		t.Errorf("to_address mismatch: got %q, want %q", decoded.ToAddress, original.ToAddress)
	}
	if decoded.Type != original.Type {
		t.Errorf("type mismatch: got %d, want %d", decoded.Type, original.Type)
	}
	if len(decoded.MessageId) != len(original.MessageId) {
		t.Errorf("message_id length mismatch: got %d, want %d", len(decoded.MessageId), len(original.MessageId))
	}
	if decoded.Timestamp != original.Timestamp {
		t.Errorf("timestamp mismatch: got %d, want %d", decoded.Timestamp, original.Timestamp)
	}

	// Verify oneof payload
	enc, ok := decoded.Payload.(*pinchv1.Envelope_Encrypted)
	if !ok {
		t.Fatal("payload is not EncryptedPayload")
	}
	if len(enc.Encrypted.Nonce) != 24 {
		t.Errorf("nonce length: got %d, want 24", len(enc.Encrypted.Nonce))
	}
	if string(enc.Encrypted.Ciphertext) != string(ciphertext) {
		t.Errorf("ciphertext mismatch")
	}
	if len(enc.Encrypted.SenderPublicKey) != 32 {
		t.Errorf("sender_public_key length: got %d, want 32", len(enc.Encrypted.SenderPublicKey))
	}
}

func TestPlaintextPayloadRoundTrip(t *testing.T) {
	original := &pinchv1.PlaintextPayload{
		Version:     1,
		Sequence:    42,
		Timestamp:   time.Now().UnixMilli(),
		Content:     []byte("hello"),
		ContentType: "text/plain",
	}

	// Serialize
	data, err := proto.Marshal(original)
	if err != nil {
		t.Fatalf("failed to marshal plaintext payload: %v", err)
	}

	// Deserialize
	decoded := &pinchv1.PlaintextPayload{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("failed to unmarshal plaintext payload: %v", err)
	}

	// Verify all fields
	if decoded.Version != 1 {
		t.Errorf("version: got %d, want 1", decoded.Version)
	}
	if decoded.Sequence != 42 {
		t.Errorf("sequence: got %d, want 42", decoded.Sequence)
	}
	if decoded.Timestamp != original.Timestamp {
		t.Errorf("timestamp mismatch: got %d, want %d", decoded.Timestamp, original.Timestamp)
	}
	if string(decoded.Content) != "hello" {
		t.Errorf("content: got %q, want %q", string(decoded.Content), "hello")
	}
	if decoded.ContentType != "text/plain" {
		t.Errorf("content_type: got %q, want %q", decoded.ContentType, "text/plain")
	}
}

func TestHandshakePayload(t *testing.T) {
	signingKey := make([]byte, 32)
	encryptionKey := make([]byte, 32)
	for i := range signingKey {
		signingKey[i] = byte(i)
		encryptionKey[i] = byte(i + 32)
	}

	original := &pinchv1.Envelope{
		Version:     1,
		FromAddress: "pinch:abc123@relay.example.com",
		Type:        pinchv1.MessageType_MESSAGE_TYPE_HANDSHAKE,
		Payload: &pinchv1.Envelope_Handshake{
			Handshake: &pinchv1.Handshake{
				Version:       1,
				SigningKey:    signingKey,
				EncryptionKey: encryptionKey,
			},
		},
	}

	data, err := proto.Marshal(original)
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}

	decoded := &pinchv1.Envelope{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	hs, ok := decoded.Payload.(*pinchv1.Envelope_Handshake)
	if !ok {
		t.Fatal("payload is not Handshake")
	}
	if hs.Handshake.Version != 1 {
		t.Errorf("handshake version: got %d, want 1", hs.Handshake.Version)
	}
	if len(hs.Handshake.SigningKey) != 32 {
		t.Errorf("signing_key length: got %d, want 32", len(hs.Handshake.SigningKey))
	}
	if len(hs.Handshake.EncryptionKey) != 32 {
		t.Errorf("encryption_key length: got %d, want 32", len(hs.Handshake.EncryptionKey))
	}
}

func TestAuthChallengePayload(t *testing.T) {
	nonce := make([]byte, 32)
	for i := range nonce {
		nonce[i] = byte(i + 11)
	}

	now := time.Now().UnixMilli()
	original := &pinchv1.Envelope{
		Version:   1,
		Type:      pinchv1.MessageType_MESSAGE_TYPE_AUTH_CHALLENGE,
		Timestamp: now,
		Payload: &pinchv1.Envelope_AuthChallenge{
			AuthChallenge: &pinchv1.AuthChallenge{
				Version:     1,
				Nonce:       nonce,
				IssuedAtMs:  now,
				ExpiresAtMs: now + 10_000,
				RelayHost:   "relay.example.com",
			},
		},
	}

	data, err := proto.Marshal(original)
	if err != nil {
		t.Fatalf("failed to marshal auth challenge: %v", err)
	}

	decoded := &pinchv1.Envelope{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("failed to unmarshal auth challenge: %v", err)
	}

	ac, ok := decoded.Payload.(*pinchv1.Envelope_AuthChallenge)
	if !ok {
		t.Fatal("payload is not AuthChallenge")
	}
	if ac.AuthChallenge.Version != 1 {
		t.Errorf("auth challenge version: got %d, want 1", ac.AuthChallenge.Version)
	}
	if len(ac.AuthChallenge.Nonce) != 32 {
		t.Errorf("nonce length: got %d, want 32", len(ac.AuthChallenge.Nonce))
	}
	if ac.AuthChallenge.IssuedAtMs != now {
		t.Errorf("issued_at_ms mismatch: got %d, want %d", ac.AuthChallenge.IssuedAtMs, now)
	}
	if ac.AuthChallenge.ExpiresAtMs != now+10_000 {
		t.Errorf("expires_at_ms mismatch: got %d, want %d", ac.AuthChallenge.ExpiresAtMs, now+10_000)
	}
	if ac.AuthChallenge.RelayHost != "relay.example.com" {
		t.Errorf("relay_host mismatch: got %q", ac.AuthChallenge.RelayHost)
	}
}

func TestAuthResponsePayload(t *testing.T) {
	pub := make([]byte, 32)
	sig := make([]byte, 64)
	nonce := make([]byte, 32)
	for i := range pub {
		pub[i] = byte(i)
		nonce[i] = byte(100 + i)
	}
	for i := range sig {
		sig[i] = byte(200 + (i % 32))
	}

	original := &pinchv1.Envelope{
		Version: 1,
		Type:    pinchv1.MessageType_MESSAGE_TYPE_AUTH_RESPONSE,
		Payload: &pinchv1.Envelope_AuthResponse{
			AuthResponse: &pinchv1.AuthResponse{
				Version:   1,
				PublicKey: pub,
				Signature: sig,
				Nonce:     nonce,
			},
		},
	}

	data, err := proto.Marshal(original)
	if err != nil {
		t.Fatalf("failed to marshal auth response: %v", err)
	}

	decoded := &pinchv1.Envelope{}
	if err := proto.Unmarshal(data, decoded); err != nil {
		t.Fatalf("failed to unmarshal auth response: %v", err)
	}

	ar, ok := decoded.Payload.(*pinchv1.Envelope_AuthResponse)
	if !ok {
		t.Fatal("payload is not AuthResponse")
	}
	if ar.AuthResponse.Version != 1 {
		t.Errorf("auth response version: got %d, want 1", ar.AuthResponse.Version)
	}
	if len(ar.AuthResponse.PublicKey) != 32 {
		t.Errorf("public_key length: got %d, want 32", len(ar.AuthResponse.PublicKey))
	}
	if len(ar.AuthResponse.Signature) != 64 {
		t.Errorf("signature length: got %d, want 64", len(ar.AuthResponse.Signature))
	}
	if len(ar.AuthResponse.Nonce) != 32 {
		t.Errorf("nonce length: got %d, want 32", len(ar.AuthResponse.Nonce))
	}
}

func TestMessageTypeEnumValues(t *testing.T) {
	tests := []struct {
		name  string
		value pinchv1.MessageType
		want  int32
	}{
		{"UNSPECIFIED", pinchv1.MessageType_MESSAGE_TYPE_UNSPECIFIED, 0},
		{"HANDSHAKE", pinchv1.MessageType_MESSAGE_TYPE_HANDSHAKE, 1},
		{"AUTH_CHALLENGE", pinchv1.MessageType_MESSAGE_TYPE_AUTH_CHALLENGE, 2},
		{"AUTH_RESPONSE", pinchv1.MessageType_MESSAGE_TYPE_AUTH_RESPONSE, 3},
		{"MESSAGE", pinchv1.MessageType_MESSAGE_TYPE_MESSAGE, 4},
		{"DELIVERY_CONFIRM", pinchv1.MessageType_MESSAGE_TYPE_DELIVERY_CONFIRM, 5},
		{"CONNECTION_REQUEST", pinchv1.MessageType_MESSAGE_TYPE_CONNECTION_REQUEST, 6},
		{"CONNECTION_RESPONSE", pinchv1.MessageType_MESSAGE_TYPE_CONNECTION_RESPONSE, 7},
		{"HEARTBEAT", pinchv1.MessageType_MESSAGE_TYPE_HEARTBEAT, 8},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if int32(tt.value) != tt.want {
				t.Errorf("MessageType %s: got %d, want %d", tt.name, int32(tt.value), tt.want)
			}
		})
	}
}
